kube-prometheus/jsonnet: Use jsonnet-bundler

brancz opened the pull request:
This PR has some large changes on the kube-prometheus sub-project. It now uses a packaging approach of jsonnet code using [jsonnet-bundler](https://github.com/jsonnet-bundler/jsonnet-bundler). This allows each component to be individually packaged up and then brought together in the now meta-package kube-prometheus. As such, Grafana has already been extracted into it's own package: [kubernetes-grafana](https://github.com/brancz/kubernetes-grafana). All other components are currently still living in this repository but the goal is to eventually make them standalone packages. Additionally the majority of alerting rules and dashboards have been extracted to the [Kubernetes monitoring mixins](https://github.com/kubernetes-monitoring/kubernetes-mixin). Not all alerting rules and dashboards have been migrated there, but I'd prefer to get this merged and continue with the migration piece by piece (which is why dashboards and alerting rules have not been entirely removed from this repository).

@mxinden could you just play through the instructions in the readme against a local minikube as described in the readme? If that works I think we can merge this.

mxinden commented on contrib/kube-prometheus/examples/kubeadm.jsonnet:
```diff
@@ -0,0 +1,2 @@
+(import "kube-prometheus/kube-prometheus.libsonnet") +
```
Why is `kubeadm.jsonnet` in `contrib/kube-prometheus/examples/` and  `kube-prometheus-minikube.jsonnet` in `contrib/kube-prometheus/hack/scripts`?

brancz replied:
minikube is a combination of kubeadm+node-ports, which is an opinionated way of running this stack, and nicely demonstrates how users may use it. Do you think we should use it in a concrete example?

brancz replied:
Added a full minikube example.

mxinden replied:
Addressed, thanks.

mxinden commented on Documentation/user-guides/cluster-monitoring.md:
```diff
@@ -37,7 +37,7 @@ The manifests used here use the [Prometheus Operator](https://github.com/coreos/
 > In future versions of minikube and kubeadm this will be the default, but for the time being, we will have to configure it ourselves.
 
-[embedmd]:# (../../contrib/kube-prometheus/manifests/prometheus-operator/prometheus-operator-deployment.yaml)
+[embedmd]:# (../../contrib/kube-prometheus/manifests/0prometheus-operator-deployment.yaml)
```
What is the leading `0` needed for in `0prometheus-operator-deployment`? For load precedence?

brancz replied:
yes, that way we can just run `kubectl create -f manifests/`, which is a very smooth getting started experience I think

mxinden commented on contrib/kube-prometheus/README.md:
```diff
@@ -2,221 +2,184 @@
+## Quickstart
 
-See `example-dist/kubeadm` for an example for deploying on minikube, using the minikube kubeadm bootstrapper. The `example-dist/kubeadm/kube-prometheus.jsonnet` file renders the kube-prometheus manifests using jsonnet and then merges the result with kubeadm specifics, such as information on how to monitor kube-controller-manager and kube-scheduler as created by kubeadm. In addition for demonstrat ...
+Although this project is intended to be used as a library, a compiled version of the Kubernetes manifests generated with this library is checked into this repository in order to try the content our quickly.
```
`s/the content our/the content out/`

mxinden commented on contrib/kube-prometheus/hack/scripts/build-jsonnet.sh:
```diff
@@ -8,18 +8,14 @@ json="tmp/manifests.json"
     path="${prefix}/${dir}"
     mkdir -p ${path}
-    jq -r ".[\"${file}\"]" ${json} | gojsontoyaml -yamltojson | gojsontoyaml > "${prefix}/${file}"
+    fullfile=$(echo ${file} | sed -r 's/([a-z0-9])([A-Z])/\1-\L\2/g' | tr '[:upper:]' '[:lower:]')
```
Or reuse the script you link in `/contrib/kube-prometheus/README.md`?

brancz replied:
Yes you're right, that's a better idea. In fact I'll just get rid of the hack directory all together, it was a bad practice to begin with.

Do you think normalizing the file name is excessive or makes sense? I feel it makes the script look less approachable and regex seems scary. What do you think?

mxinden replied:
Thanks.

I think once it has a description above, the regex is not that scary anymore. I like the consistency across filenames. Your call. No strong opinion.

brancz replied:
I went without regexing now, the build script is quite simple this way. Let me know what you think.

mxinden commented on contrib/kube-prometheus/README.md:
```diff
@@ -2,221 +2,184 @@
-For the `kube-scheduler` and `kube-controller-manager` there are headless
-services prepared, simply add them to your running cluster:
+kubeadm:
+[embedmd]:# (examples/kubeadm.jsonnet)
```
This is not properly expanded.

mxinden commented on contrib/kube-prometheus/README.md:
````diff
@@ -2,221 +2,184 @@
-kubectl -n kube-system create -f manifests/k8s/
-```
+bootkube:
+[embedmd]:# (examples/bootkube.jsonnet)
````
This is not properly expanded.

mxinden commented on contrib/kube-prometheus/README.md:
```diff
@@ -2,221 +2,184 @@
-working cluster, but is typically deployed outside of it. This monitoring
-setup assumes that it is made visible from within the cluster through a headless
-service as well.
+[embedmd]:# (examples/node-ports.jsonnet)
```
This is not properly expanded.

mxinden commented on contrib/kube-prometheus/README.md:
````diff
@@ -2,221 +2,184 @@
 
 ```
-minikube delete && minikube start --kubernetes-version=v1.9.6 --memory=4096 --bootstrapper=kubeadm --extra-config=kubelet.authentication-token-webhook=true --extra-config=kubelet.authorization-mode=Webhook --extra-config=scheduler.address=0.0.0.0 --extra-config=controller-manager.address=0.0.0.0
+$ kubectl create -f manifests/
````
Not sure with the discussion on [twitter](https://twitter.com/fredbrancz/status/990691586836193282) we can really rely on all CRDs being in place in time, right?

brancz replied:
if it works consistently for us, I think we should stick with it, as it provides the smoothest getting started experience in my opinion

mxinden commented on contrib/kube-prometheus/README.md:
````diff
@@ -2,221 +2,184 @@
 
 ```
-minikube delete && minikube start --kubernetes-version=v1.9.6 --memory=4096 --bootstrapper=kubeadm --extra-config=kubelet.authentication-token-webhook=true --extra-config=kubelet.authorization-mode=Webhook --extra-config=scheduler.address=0.0.0.0 --extra-config=controller-manager.address=0.0.0.0
+$ kubectl create -f manifests/
````
`/manifests` seems to miss the monitoring namespace manifest.

brancz replied:
do you think the namespace object should be part of the stack or we should just preceed this command with `kubectl create namespace monitoring`?

mxinden replied:
To me the `monitoring` namespace definition should be part of the kube-prometheus project. I would prefer the declarative way via a manifest instead of the imperative way via `kubectl`.

brancz replied:
Added namespace object.

ant31 replied:
Is there a dependency to run inside 'monitoring'?

brancz commented:
Tested latest head and addressed all comments. Could you give this a re-review @mxinden ? :slightly_smiling_face: (I put all fixes in a separate commit so it's better reviewable, ultimately I would like to squash the two though when we merge)
